#! /bin/sh -e

# All lines beginning with `# DP:' are a description of the patch.
# DP: Description: Fix segfault in getaddrinfo
# DP: Author: Ulrich Drepper  <drepper@redhat.com>
# DP: Upstream status: In CVS
# DP: Date: December 24, 2002

if [ $# -ne 2 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch) patch -d "$2" -f --no-backup-if-mismatch -p1 < $0;;
    -unpatch) patch -d "$2" -f --no-backup-if-mismatch -R -p1 < $0;;
    *)
	echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
	exit 1
esac
exit 0

# append the patch here and adjust the -p? flag in the patch calls.
2002-12-16  Ulrich Drepper  <drepper@redhat.com>

	* sysdeps/posix/getaddrinfo.c (gaih_inet): If __nss_lookup_function
	fails to return a function pointer don't use it.

--- glibc-2.3.1.orig/sysdeps/posix/getaddrinfo.c	2002-12-23 10:47:34.000000000 -0500
+++ glibc-2.3.1/sysdeps/posix/getaddrinfo.c	2002-12-24 10:08:27.000000000 -0500
@@ -570,25 +570,28 @@ gaih_inet (const char *name, const struc
 		{
 		  fct = __nss_lookup_function (nip, "gethostbyname2_r");
 
-		  gethosts2 (AF_INET6, struct in6_addr);
-		  no_inet6_data = no_data;
-		  inet6_status = status;
-		  gethosts2 (AF_INET, struct in_addr);
-
-		  /* If we found one address for AF_INET or AF_INET6,
-		     don't continue the search.  */
-		  if (inet6_status == NSS_STATUS_SUCCESS ||
-		      status == NSS_STATUS_SUCCESS)
-		    break;
-
-		  /* We can have different states for AF_INET
-		     and AF_INET6. Try to find a usefull one for
-		     both.  */
-		  if (inet6_status == NSS_STATUS_TRYAGAIN)
-		    status = NSS_STATUS_TRYAGAIN;
-		  else if (status == NSS_STATUS_UNAVAIL &&
-			   inet6_status != NSS_STATUS_UNAVAIL)
-		    status = inet6_status;
+		  if (fct != NULL)
+		    {
+		      gethosts2 (AF_INET6, struct in6_addr);
+		      no_inet6_data = no_data;
+		      inet6_status = status;
+		      gethosts2 (AF_INET, struct in_addr);
+
+		      /* If we found one address for AF_INET or AF_INET6,
+			 don't continue the search.  */
+		      if (inet6_status == NSS_STATUS_SUCCESS ||
+			  status == NSS_STATUS_SUCCESS)
+			break;
+
+		      /* We can have different states for AF_INET
+			 and AF_INET6. Try to find a usefull one for
+			 both.  */
+		      if (inet6_status == NSS_STATUS_TRYAGAIN)
+			status = NSS_STATUS_TRYAGAIN;
+		      else if (status == NSS_STATUS_UNAVAIL &&
+			       inet6_status != NSS_STATUS_UNAVAIL)
+			status = inet6_status;
+		    }
 
 		  if (nss_next_action (nip, status) == NSS_ACTION_RETURN)
 		    break;
